workflow/scripts/plot-assembly-comparison.py

```diff
@@ -0,0 +1,164 @@
+import sys
+from altair.vegalite.v4.schema.channels import Opacity
+import pandas as pd
+import pysam
+import altair as alt
+
+sys.stderr = open(snakemake.log[0], "w")
+
+data = pd.DataFrame()
+
+
+def register_lengths(sample, file_list, state, amplicon_state, data):
+    for file, assembler in zip(file_list, snakemake.params.assembler):
+        if state in ("initial", "final"):
+            with pysam.FastxFile(file) as infile:
+
+                data = data.append(
+                    {
+                        "Sample": sample,
+                        "Assembler": assembler,
+                        "Amplicon": amplicon_state,
+                        "length (bp)": max(len(contig.sequence) for contig in infile),
+                        "State": state,
+                    },
+                    ignore_index=True,
+                )
+        else:
+            quastDf = pd.read_csv(file, sep="\t")
+            data = data.append(
+                {
+                    "Sample": sample,
+                    "Assembler": assembler,
+                    "Amplicon": amplicon_state,
+                    "length (bp)": quastDf.loc[0, "N50"],
+                    "State": "N50",
+                    "Genome fraction (%)": quastDf.loc[0, "Genome fraction (%)"]
+                    if "Genome fraction (%)" in quastDf.columns
+                    else float("nan"),
+                },
+                ignore_index=True,
+            )
+    return data
+
+
+for sample, amplicon_state in zip(
+    snakemake.params.samples, snakemake.params.amplicon_state
+):
+    data = register_lengths(
+        sample,
+        [x for x in snakemake.input.initial if sample in x],
+        "initial",
+        amplicon_state,
+        data,
+    )
+    data = register_lengths(
+        sample,
+        [x for x in snakemake.input.final if sample in x],
+        "final",
+        amplicon_state,
+        data,
+    )
+    data = register_lengths(
+        sample,
+        [x for x in snakemake.input.quast if sample in x],
+        "N50",
+        amplicon_state,
+        data,
+    )
```

Comment:
code duplication, turn into a function and run 3 times with different parameters instead.